pkg/prober/prober.go

```diff
@@ -0,0 +1,155 @@
+package prober
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/common/route"
+)
+
+const (
+	healthyEndpointPath  = "/-/healthy"
+	readyEndpointPath    = "/-/ready"
+	okProbeText          = "thanos %v is %v"
+	errorProbeText       = "thanos %v is not %v. Reason: %v"
+	probeErrorHTTPStatus = 500
+	initialErrorText     = "thanos %s is initializing"
+)
+
+// Prober represents health and readriness status of given compoent.
+type Prober struct {
+	logger       log.Logger
+	loggerMtx    sync.Mutex
+	componentMtx sync.Mutex
+	component    string
+	readyMtx     sync.Mutex
+	readiness    error
+	healthyMtx   sync.Mutex
+	healthiness  error
+}
+
+// SetLogger sets logger used by the Prober.
+func (p *Prober) SetLogger(logger log.Logger) {
+	p.loggerMtx.Lock()
+	defer p.loggerMtx.Unlock()
+	p.logger = logger
+}
+
+func (p *Prober) getLogger() log.Logger {
+	p.loggerMtx.Lock()
+	defer p.loggerMtx.Unlock()
+	return p.logger
+}
+
+// SetComponent sets component name of the Prober displayed in responses.
+func (p *Prober) SetComponent(component string) {
+	p.componentMtx.Lock()
+	defer p.componentMtx.Unlock()
+	p.component = component
+}
+
+func (p *Prober) getComponent() string {
+	p.componentMtx.Lock()
+	defer p.componentMtx.Unlock()
+	return p.component
+}
+
+// NewProber returns Prober reprezenting readiness and healthiness of given component.
+func NewProber(component string, logger log.Logger) *Prober {
+	initialErr := fmt.Errorf(initialErrorText, component)
+	prober := &Prober{}
+	prober.SetComponent(component)
+	prober.SetLogger(logger)
+	prober.SetNotHealthy(initialErr)
+	prober.SetNotReady(initialErr)
+	return prober
+}
+
+// HandleInMux registers readiness and liveness probes to mux.
+func (p *Prober) RegisterInRouter(router *route.Router) {
```

Comment:
I feel like if we used `route.Router` everywhere we could rename this to `Register` and drop the mux support. Is there any specific reason we are using both? @bwplotka  wdyt?

Reply:
No. The only reason is `api` packages and `ui` we copied from Prometheus and inconsistency was there (or maybe rather we used mux and they router). Happy to move to single one at some point.

Reply:
Ok, so should I unify this in this PR or is it out of scope? And if yes which one will it be?

Reply:
I think it's out of scope